src/ast/ast_validations.c

```diff
@@ -1123,30 +1123,16 @@ static void _AST_GetReferredIdentifiers(const cypher_astnode_t *node, rax *ident
 	}
 }
 
-/* Check that all referred identifiers been defined in the same AST scope. */
-static AST_Validation _Validate_Aliases_DefinedInScope(const AST *ast, uint start_offset,
-													   uint end_offset) {
+static AST_Validation _Validate_Aliases_DefinedInClause(const cypher_astnode_t *clause,
+														rax *defined_aliases) {
 	AST_Validation res = AST_VALID;
-	rax *defined_aliases = raxNew();
 	rax *referred_identifiers = raxNew();
 
-	for(uint i = start_offset; i < end_offset; i ++) {
-		const cypher_astnode_t *clause = cypher_ast_query_get_clause(ast->root, i);
-		if(cypher_astnode_type(clause) == CYPHER_AST_WITH) {
-			/* If this is a WITH clause, we only want to collect defined aliases if this is the start
-			 * of the segment, and only want to collect referred aliases if this is the end of the segment.
-			 * Otherwise, queries like "MATCH (a) WITH e RETURN e" would incorrectly register 'e' as a valid reference. */
-			if(i == start_offset) _AST_GetDefinedIdentifiers(clause, defined_aliases);
-			else if(i == end_offset - 1) _AST_GetReferredIdentifiers(clause, referred_identifiers);
-			continue;
-		}
+	// Get defined identifiers.
+	_AST_GetDefinedIdentifiers(clause, defined_aliases);
```

Comment:
What happens in the case where a clause tries to redefined an already defined alias?

Reply:
Right now nothing happens, since in many cases this would be legal. For example, MATCH and MERGE would both currently add `n` to the definition map in a query like:
```
MATCH (n {val: 1}) MERGE (n)-[:E]->(m)
```